include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,648 @@
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_DECODED_VIEW_HPP
+#define BOOST_URL_DECODED_VIEW_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <boost/url/string_view.hpp>
+#include <boost/url/pct_encoding.hpp>
+#include <type_traits>
+#include <iterator>
+#include <iosfwd>
+
+namespace boost {
+namespace urls {
+
+/** A view of the percent decoded characters in a string
+
+    This class is a lightweight wrapper that presents
+    a view of an underlying sequence of percent-encoded
+    characters mutating or copying it.
+
+    The reference objects do not own the underlying
+    character buffer, are inexpensive to create and
+    copy. This avoid any unnecessary memory allocations
+    when working with decoded strings, and transfers
+    this responsibility to the user.
+
+    The decoded_view iterators operate lazily on
+    underlying sequence transforming groups
+    of percent-encoded characters into single
+    decoded elements.
+
+    The underlying encoded data is constant and
+    is never mutated by the view. Thus, the views
+    only contain the const-qualified functions
+    `begin()`/`end()`.
+
+    Any operation on the underlying percent-encoded
+    strings invalidates the view iterators. The view
+    should be recreated after any operation that
+    may mutate the underlying range.
+
+*/
+class decoded_view
+{
+    char const* p_{nullptr};
+    std::size_t n_{0};
+    std::size_t dn_{0};
+    bool plus_to_space_{true};
+
+    using traits_type = std::char_traits<char>;
+
+public:
+    /** Type of a decoded character
+    */
+    using value_type = char;
+
+    /// @copydoc value_type
+    using reference = char;
+
+    /// @copydoc value_type
+    using const_reference = char;
+
+    /** The unsigned integer type used to represent size.
+    */
+    using size_type = std::size_t;
+
+    /** The signed integer type used to represent differences.
+    */
+    using difference_type = std::ptrdiff_t;
+
+    /** A read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+    */
+#ifdef BOOST_URL_DOCS
+    using iterator = __see_below__;
+#else
+    class iterator;
+#endif
+
+    /// @copydoc const_iterator
+    using const_iterator = iterator;
+
+    /** A reverse read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+    */
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    /// @copydoc reverse_iterator
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+    */
+    decoded_view(
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(opt.plus_to_space)
```

Comment:
if you already have in-class initializers what's the purpose of repeating it in the ctor-init list?